# Hemiplasy hypothesis - inference-based approach


*environiments:* yaml.main 


*aim:* infer species trees / infer gene concordance factors (gCFs) / run HeIST

As input for HeIST, it is necessary to prepare a txt file including two species trees, one of which indicates the gene concordance factors (gCFs), a measure for quantifying genealogical concordance in phylogenomic datasets. For every branch of a reference tree, gCF is defined as the percentage of “decisive” gene trees containing that branch. 

**NB:** In order to test the performances of subsequent analyses on different subsets of data, both sets of orthogroups are split into two subsets: one including all orthogroups (with a minimum of 5 species each) and one including only orthogroups comprising all 10 species. 

Each of the following commands must thus be applied to four subsets, in total: 
* all aminoacids-composed orthogroups (min 5 species);
* aminoacids-composed orthogroups with minimimum 10 species;
* all nucleotides-composed orthogroups (min 5 species);
* nucleotides-composed orthogroups with minimimum 10 species.

## Infer species tree for both nucleotidic and aminoacidic sequences


**NB:** The species trees are inferred using this [script for aminoacidic sequences](https://github.com/MattiaRag/timemaproject/blob/main/scripts/species_tree_aa.sh) and this [script for nucleotidic sequences](https://github.com/MattiaRag/timemaproject/blob/main/scripts/species_tree_nu.sh). The current chapter provides a detailed description of commands constituting the concerned pipeline, taking aminoacidic sequences as example model. 


At first, all trimmed files are copied and paste into the new directory `sptree_aa/min_5sp`:


```
cp preliminary/orthogroupsdisco_aa/Trimmed_aa/*.fa species_trees/sptree_aa/min_5sp
```

One further directory called `sptree_aa/min_10sp` is generated and orthogroups' fastas including all 10 species are copied and paste in there. These steps are performed through a `for` cycle:

```
for i in species_trees/sptree_aa/min_5sp/*.fa; do
    # Count the number of headers in the file
    num_headers=$(grep -c '^>' "$i")

    # If the number of headers is greater than or equal to 10
    if [ "$num_headers" -ge 10 ]; then
        # Copy the file to min_10sp directory
        cp "$i" species_trees/sptree_aa/min_10sp/
    else
        echo "The file $i contains less than 10 headers. Skipping to the next step."
    fi
done
```

The following steps must be performed on both `min_5sp` and `min_10sp` directories.

Before concatenation, the fastas' headers are renamed keeping just the species names (e.g. TPA, TBI, etc.), while empty files are removed: 

```
awk '/^>/{split(species_trees/sptree_aa/min_5sp, arr, "_"); print arr[1]; next} {print}' "orth.fa" > species_trees/sptree_aa/min_5sp/Renamed/orth.rename
find species_trees/sptree_aa/min_5sp/Renamed -type f -size 0 -delete
```

Concatenation is then operated using AMAS.py on the `Renamed` directory:

```
python scripts/AMAS.py concat -y nexus -i species_trees/sptree_aa/min_5sp/Renamed/*.rename -f fasta -d aa -t species_trees/sptree_aa/min_5sp/Iqtreeinput/concat_file
```

The flag `-y` specifies the format of the output file. 
The flag `-d` is used to specify the type of data being processed. This flag takes value `aa` for aminoacidic and `dna` for nucleotidic sequences.
The flag `-f` is used to specify the format of the input files.

Concatenation is directly followed by IQ-TREE:

```
iqtree2 -m TESTNEW -bb 1000 -s species_trees/sptree_aa/min_5sp/Iqtreeinput/concat_file --prefix species_trees/sptree_aa/min_5sp/Iqtreeoutput/concat_file.faa -nt AUTO
```

The flag `-bb` enables ultrafast bootstrap analysis with 1000 bootstrap replicates.
The flag `--prefix` sets the prefix for all output files generated by IQ-TREE.
The flag `-m TESTNEW` specifies the model selection strategy, involving a comprehensive model selection method that tests all substitution models and chooses the best one.

The computed species tree can be retrieved [here](https://github.com/MattiaRag/timemaproject/tree/main/intermediate_files/speciestree).

## Infer gene concordance factors (gCFs)


Once inferred the gene trees and species trees with both aminoacidic and nucleotidic sequences, it is possible to infer the gene concordance factors. 

**NB:** all commands needed for obtaining gCFs have been implemented within this [script](https://github.com/MattiaRag/timemaproject/blob/main/scripts/gCF.sh).

The concordance factors can be easily inferred through a single command line:

```
iqtree2 -t concat.treefile --gcf loci.treefile --prefix concord
```

The flag `-t` specifies the species tree.
The flag `--gcf` specifies a `loci.treefile`, a file including the list of gene trees in newick format. 

All four inferred species trees and related gCFs are represented in [Fig. S1](https://github.com/MattiaRag/timemaproject/blob/main/pictures/Fig_S1.png), while the species tree derived from aminoacidic sequences of minimum 5 species-orthogroups is represented in [Fig. 1](https://github.com/MattiaRag/timemaproject/blob/main/pictures/Fig_1.png) as 1A.

## Run HeIST

#### Prepare the NEXUS input for HeIST

The input file for HeIST is a modified NEXUS format. It includes two species trees (in newick format) and a list of taxa showcasing the derived character to investigate (parthenogenesis, in this case).  

The first species tree (must be named `tree_1`) should have branch lengths in average substitutions per site and branches must be labeled with concordance factors. The second tree (named `tree_2`) should have the same branch lengths as tree_1, but have internal branch labels rather than concordance factors. These labels can be arbitrary (I1-I8, in this case). The computed input files can be retrieved [here](https://github.com/MattiaRag/timemaproject/tree/main/intermediate_files/heist_inputs).

```
#NEXUS
begin trees;
tree tree_1 = (TBI:0.0216811821,(((TCE:0.0258830720,TMS:0.0223835875)85.6:0.0313337511,((TCM:0.0257016099,TSI:0.0215270659)75:0.0211077249,(TDI:0.0227873921,TPS:0.0298985030)68.4:0.0167929055)73:0.0234755438)83.9:0.0502147658,(TGE:0.0282544807,TPA:0.0490445735)60.1:0.0097146123)79.1:0.0203223997,TTE:0.0170368202);
tree tree_2 = (TBI:0.0216811821,(((TCE:0.0258830720,TMS:0.0223835875)I1:0.0313337511,((TCM:0.0257016099,TSI:0.0215270659)I2:0.0211077249,(TDI:0.0227873921,TPS:0.0298985030)I3:0.0167929055)I4:0.0234755438)I5:0.0502147658,(TGE:0.0282544807,TPA:0.0490445735)I6:0.0097146123)I7:0.0203223997,TTE:0.0170368202)I8;
end;

begin hemiplasytool;
set derived taxon=TDI
set derived taxon=TSI
set derived taxon=TMS
set derived taxon=TGE
set derived taxon=TTE
set conversion type=extend
end;
```

#### Run HeIST


HeIST can be run through the following command line:

```
heist -v -n 100000000 -t 12 -p ../../../HeIST/lib/msdir/ms -g ../../../HeIST/lib/Seq-Gen-1.3.4/source/seq-gen -o ./output_10_aa_0.0005.txt ./heistinput_aa_min10.txt -s 0.0005
```

The flag `-v` (or `--verbose`) enables debugging messages to be displayed.
The flag `-n` specifies the number of replicates per batch.
The flag `-p` specifies the path to ms dependency.
The flag `-g` specifies the path to seq-gen dependency.
The flag `-o` indicates the output directory/prefix.
The flag `-s` specifies the Seq-gen mutation rate.

Four different values of Seq-gen mutation rate `-s` are tested: `0.01`, `0.005`, `0.001`, `0.0005`. The number of focal cases detected by HeIST will be of course lower the lower the mutation rate. The values' range was calculated according to the maximum, average and minimum known values of population-scaled mutation rate. A lower value (`0.0005`), compared to the minimum of `0.001`, was adopted as HeIST assumes that transitions between character states are controlled by a single site, and therefore that the nucleotide mutation rate is a good approximation of the trait mutation rate. 

As the inferential approach adopted by this tool is computationally demanding, we used to run the command line multiple times on the same input file, just to later merge the output through a properly designed [script](https://github.com/MattiaRag/timemaproject/blob/main/scripts/heist_summary.py), summarizing data included in the `RESULTS` section of each `.txt` HeIST output file, for both aminoacids and nucleotides:

```
python summary.py heist_aa
python summary.py heist_nu
```

This script accepts, as input, the directories including all sub-directories composed of results from several runs of the same HeIST analyses, characterized by a certain combination of parameters. The result is a directory, for both [aminoacids](https://github.com/MattiaRag/timemaproject/tree/main/intermediate_files/heist_aa_summary) and [nucleotides](https://github.com/MattiaRag/timemaproject/tree/main/intermediate_files/heist_nu_summary), including all resulting `.txt` files. 

The data on number of focal cases recognized as representing "Full hemiplasy", "Some hemiplasy" and "Full homoplasy" were subsequently included into a `.tsv` files for both [aminoacids](https://github.com/MattiaRag/timemaproject/blob/main/intermediate_files/heist_aa_summary.tsv) and [nucleotides](https://github.com/MattiaRag/timemaproject/blob/main/intermediate_files/heist_nu_summary.tsv), using a proper [script](https://github.com/MattiaRag/timemaproject/blob/main/scripts/heist_summary_extract.py):

```
python heist_summary_extract.py heist_aa_summary heist_aa_summary.tsv
python heist_summary_extract.py heist_nu_summary heist_nu_summary.tsv
```

This script accepts, as input, the output files produced by `summary.py`.

### Resulting graphs


A proper [R script](https://github.com/MattiaRag/timemaproject/blob/main/scripts/Rscripts/part_1.R.R) was used to represent the frequency of hemiplasy across simulations and empirically ([Fig. 2A](https://github.com/MattiaRag/timemaproject/blob/main/pictures/2A.pdf)), as well as a HeIST sensitivity analyses leveraging the species tree and the corresponding gene concordance factors (gCFs) ([Fig. S4](https://github.com/MattiaRag/timemaproject/blob/main/pictures/S4.pdf)).
The script accepts as input the following files:
* [`heist_aa_summary.tsv`](https://github.com/MattiaRag/timemaproject/blob/main/intermediate_files/heist_aa_summary.tsv);
* [`heist_nu_summary.tsv`](https://github.com/MattiaRag/timemaproject/blob/main/intermediate_files/heist_nu_summary.tsv);
* [`fitch_results_aa.txt`](https://github.com/MattiaRag/timemaproject/blob/main/intermediate_files/fitch_results/fitch_results_aa.txt);
* [`fitch_results_nu.txt`](https://github.com/MattiaRag/timemaproject/blob/main/intermediate_files/fitch_results/fitch_results_nu.txt);
* [`fitch_asr_input_aa.txt`](https://github.com/MattiaRag/timemaproject/blob/main/intermediate_files/fitch_asr_inputs/fitch_asr_input_aa.txt);
* [`fitch_asr_input_nu.txt`](https://github.com/MattiaRag/timemaproject/blob/main/intermediate_files/fitch_asr_inputs/fitch_asr_input_nu.txt).


<div style="text-align: center;">
  <figure style="display: inline-block; text-align: center; margin: 0;">
    <img src="https://github.com/MattiaRag/timemaproject/blob/main/pictures/2A-1.png?raw=true" alt="Fig. 2A" width="600">
    <figcaption style="margin-top: 10px;">
      <sub><strong>Fig. 2A:</strong> Frequency of hemiplasy across simulations and empirically. The first four bar plots represent the frequency of full and partial hemiplasy in the focal cases simulated using HeIST at four different mutation rates; the numbers indicate the count of focal cases that support each scenario. The analysis is based on the gene concordance factors inferred using the 10867 amino acid alignments of single-copy genes present in more than five Timema species that were used to infer the species tree in Figure 1. The results using nucleotide alignments or nucleotide and amino acids alignments of single-copy genes present in at least five Timema species can be found in Figure S3. The last bar plot represents the different homoplastic/hemiplasic scenarios calculated using Fitch parsimony on 3,813 genes found in at least four parthenogenetic and four gonochoric species and no branches shorter than 1e-6; numbers indicate the count of gene trees supporting each scenario.</sub>
    </figcaption>
  </figure>
</div>



---


[main](https://github.com/MattiaRag/timemaproject/tree/main) /
[exp_des](https://github.com/MattiaRag/timemaproject/blob/main/markdowns/exp_design.md) /
[prel](https://github.com/MattiaRag/timemaproject/blob/main/markdowns/preliminary.md) /
1a /
[1b](https://github.com/MattiaRag/timemaproject/blob/main/markdowns/part_1b.md) /
[2a](https://github.com/MattiaRag/timemaproject/blob/main/markdowns/part_2a.md) /
[2b](https://github.com/MattiaRag/timemaproject/blob/main/markdowns/part_2b.md)  
